modules/programs/waybar.nix

```diff
@@ -0,0 +1,371 @@
+      type = nullOr str;
+      default = null;
+      description = ''
+        CSS style of the bar.
+        See <link xlink:href="https://github.com/Alexays/Waybar/wiki/Configuration/>"/> for the documentation.
+      '';
+      example = ''
+        * {
+          border: none;
+          border-radius: 0;
+          font-family: Source Code Pro;
+        }
+        window#waybar {
+          background: #16191C;
+          color: #AAB2BF;
+        }
+        #workspaces button {
+          padding: 0 5px;
+        }
+      '';
+    };
+  };
+
+  config = let
+    # Inspired by https://github.com/NixOS/nixpkgs/pull/89781
+    writePrettyJSON = name: x:
+      pkgs.runCommandNoCCLocal name { } ''
+        echo '${builtins.toJSON x}' | ${pkgs.jq}/bin/jq . > $out
+      '';
+
+    configSource = let
+      # Removes nulls because Waybar ignores them for most values
+      removeNulls = filterAttrs (_: v: !isNull v);
+      # Makes the actual valid configuration Waybar accepts (strips our custom settings before converting to JSON)
+      makeConfiguration = configuration:
+        let
+          # The "modules" option is not valid in the JSON as its descendants have to live at the top-level
+          settingsWithoutModules =
+            filterAttrs (n: _: n != "modules") configuration;
+          settingsModules =
+            optionalAttrs (configuration.modules != { }) configuration.modules;
+        in removeNulls (settingsWithoutModules // settingsModules);
+      # The clean list of configurations
+      finalConfiguration = map makeConfiguration cfg.settings;
+    in writePrettyJSON "waybar-config.json" finalConfiguration;
+
+    warnings = let
+      mkPath = idx:
+        let i = toString idx;
+        in "${modulesPath}[definition ${i}-entry ${i}]";
+      mkUnreferencedModuleWarning = idx: name:
+        "The module '${name}' defined in '${mkPath idx}' is not referenced "
+        + "in either `modules-left`, `modules-center` or `modules-right` of Waybar's options";
+      mkUndefinedModuleWarning = idx: name:
+        "The module '${name}' defined in '${mkPath idx}' is neither "
+        + "a default module or a custom module declared in '${
+          mkPath idx
+        }.modules'";
+
+      # Find all modules in `modules-{left,center,right}` and `modules` not declared/referenced.
+      # cfg.settings is a list of Waybar configurations, we need to preserve the index for appropriate warnings
+      allFaultyModules = flip imap1 cfg.settings (idx: settings:
+        let
+          allModules = concatMap (x: attrByPath [ x ] [ ] settings) [
+            "left"
+            "center"
+            "right"
+          ];
+          nonDefaultModules = subtractLists defaultModuleNames allModules;
+          declaredModules = attrNames settings.modules;
+          # Modules declared in `modules` but not referenced in `modules-{left,center,right}`
+          unreferencedModules = subtractLists nonDefaultModules declaredModules;
```

Comment:
The logic here for this warning needs a correction, my config has several declared modules that are _default_ modules, and subtracting only the list of _non-default_ modules leaves all my configured modules generating warnings about being unreferenced.

Seems like it should be subtracting `allModules`.

Reply:
You are right, this is fixed in commit 1000ea5 and further improved in commit 75b7899.

I use the following configuration to test errors:

``` nix
{ config, lib, pkgs, ... }:

{
  programs.waybar = {
    enable = true;
    systemd = {
      enable = true;
      withSwayIntegration = true;
    };
    settings = [{
      layer = "top";
      position = "top";
      height = 26;
      output = [ "DP-1" "eDP-1" "HEADLESS-1" ];
      modules-left = [ "sway/workspaces" "sway/mode" "custom/undefined-two-places" ];
      modules-center = [ "sway/window" "custom/undefined-two-places" ];
      modules-right = [
        "pulseaudio"
        "network"
        "cpu"
        "memory"
        "backlight"
        "tray"
        "battery"
        "clock"
        "custom/undefined-module"
        "invalid-custom"
      ];

      modules = {
        "sway/workspaces".all-outputs = true;
        "custom/unreferenced-1".whatever = false;
        "invalid-custom".whatever = false;
        "sway/mode" = { tooltip = false; };
        "sway/window" = { max-length = 120; };
        "idle_inhibitor" = { format = "{icon}"; };
      };
    }];
  };
}
```

Which produces the following warnings:

``` text
trace: warning: The module 'custom/undefined-two-places' defined in 'programs.waybar.settings.[].modules-left' and 'programs.waybar.settings.[].modules-center' is neither a default module or a custom module declared in 'programs.waybar.settings.[].modules'
trace: warning: The module 'custom/undefined-module' defined in 'programs.waybar.settings.[].modules-right' is neither a default module or a custom module declared in 'programs.waybar.settings.[].modules'
trace: warning: The module 'custom/unreferenced-1' defined in 'programs.waybar.settings.[].modules' is not referenced in either `modules-left`, `modules-center` or `modules-right` of Waybar's options
trace: warning: The module 'idle_inhibitor' defined in 'programs.waybar.settings.[].modules' is not referenced in either `modules-left`, `modules-center` or `modules-right` of Waybar's options
trace: warning: The custom module 'invalid-custom' defined in 'programs.waybar.settings.[].modules' is not a valid module name. A custom module's name must start with 'custom/' like 'custom/mymodule' for instance
```